### PR TITLE
Credit process update

### PR DESCRIPTION
**Pull request by stepanGerasimovich**

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,12 @@
+{
+  "compilerOptions": {
+    "lib": ["es2017"],
+    "target": "es6",
```
Comment (vladbronishevsky) [inline tsconfig.json:4]: why not es5?
ES6 should be okay for the backend. I'm just curious.

Comment (stepanGerasimovich) [reply tsconfig.json:4]: Exactly because ES6 should be fine, node supports it.
I see no reason for downgrade.

File: src/interface/Rule.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import { ConditionTest } from '../enum/ConditionTest';
+import { RuleType } from '../enum/RuleType';
+import { VariableInputType } from '../enum/VariableInputType';
+
+export interface Rule {
+  condition_test: ConditionTest;
+  rule_name: string;
+  rule_type: RuleType;
+  value_comparison_type: VariableInputType;
+  value_comparison?: string | number;
```
Comment (vladbronishevsky) [inline src/interface/Rule.ts:10]: Move to separate internal type ('string | number')

Comment (vladbronishevsky) [reply src/interface/Rule.ts:10]: It's up to you, actually.

File: src/interface/Organization.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+export interface Organization {
+  _id: string;
+  association: {
+    users: any[];
```
Comment (vladbronishevsky) [inline src/interface/Organization.ts:4]: Is there any reason to add typings for users and clients?

File: src/interface/Module.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+  lookup_name: string;
+  description?: string;
+}
+
+export interface DataIntegrationModule extends CommonModule {
+  dataintegration: DataIntegration;
+  credentials: Record<string, unknown>;
+}
+
+export type Module = DataIntegrationModule | CommonModule;
```
Comment (vladbronishevsky) [inline src/interface/Module.ts:20]: looks a bit weird.

File: src/interface/CreditProcessState.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  assignments?: Record<string, AssignmentsStateSegment>;
+  calculations?: Record<string, CalculationsStateSegment>;
+  dataintegrations?: Record<string, DataIntegrationStateSegment>;
+  datasources?: DataIntegrationDataSource[];
+  decline_reasons?: string[];
+  error?: { message: string };
+  output?: Record<string, OutputStateSegment>;
+  requirements?: Record<string, RequirementsStateSegment>;
+  scorecard?: Record<string, ScorecardStateSegment>;
+}
```
Comment (vladbronishevsky) [inline src/interface/CreditProcessState.ts:32]: new line?

File: src/helper/compareHelper.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    );
+  }
+
+  const compareTo = getVariableValue(value_comparison, value_comparison_type, state);
+
+  return compareOperands[compareOperand](variableValue, compareTo);
+};
+
+const getVariableValue = (
+  value: string | number | boolean | null | undefined,
```
Comment (vladbronishevsky) [inline src/helper/compareHelper.ts:40]: move to a separate type 'string | number | boolean | null | undefined'
or use any ?

File: src/enum/RuleType.ts [hunk 7]
```diff
@@ -0,0 +1,5 @@
+export enum RuleType {
+  Simple = 'simple',
```
Comment (vladbronishevsky) [inline src/enum/RuleType.ts:2]: While all enums has different approaches for cases?

File: src/StateManager.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+  [ModuleType.scorecard]: ScoreCardStateTransformer,
+};
+
+export default class StateManager {
+  static updateStateWithModuleResults = async (
+    module: Module,
+    segments: StateSegment[],
+    state: CreditProcessState,
+  ) => {
+    const StateTransformerClass = moduleTypeToStateTransformerClass[module.type];
```
Comment (vladbronishevsky) [inline src/StateManager.ts:34]: this method looks weird.
It creates a new instance and it is static.

Also, it changes incoming params.

File: src/helper/conditions-check/index.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
+import { ConditionsChecker } from './ConditionsChecker';
+import { CreditProcessState } from '../../interface/CreditProcessState';
+import { StrategySegment } from '../../interface/StrategySegment';
+
+export default (segments: StrategySegment[]) => {
+  const conditionsChecker = new ConditionsChecker();
+
+  return (state: CreditProcessState) => {
```
Comment (vladbronishevsky) [inline src/helper/conditions-check/index.ts:8]: how the usage looks like then ?

import checker from '../checker';

checker(p1)(p2);

File: src/index.ts [hunk 10]
```diff
@@ -0,0 +1,12 @@
+import { CompiledStrategy } from './interface/CompiledStrategy';
+import CreditPipeline from './CreditPipeline';
+
+export default class CreditProcess {
+  static createPipeline = async (compiledStrategy: CompiledStrategy) => {
```
Comment (vladbronishevsky) [inline src/index.ts:5]: let's discuss it too

File: src/module/abstract/AbstractStateTransformer.ts [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    return this.singleSegmentTransform(module, segments[0], _state);
+  }
+
+  protected async noSegmentTransform(
+    module: Module,
+    state: CreditProcessState,
+  ): Promise<CreditProcessState> {
+    state.credit_process.push({
```
Comment (vladbronishevsky) [inline src/module/abstract/AbstractStateTransformer.ts:29]: what about storing the state in a class instance?

**Review by vladbronishevsky — APPROVED** [review]

**Pull request merged**